### PR TITLE
Added implicit predicate domain support for type index actor

### PR DESCRIPTION
**Pull request by Sindhu-Vasireddy**

### CODE REVIEW

**Review by rubensworks — CHANGES_REQUESTED** [review]
Thanks for this @Sindhu-Vasireddy!
You can find my suggestions below.

**Comment by Sindhu-Vasireddy** [comment]
Hi @rubensworks, I've now addressed all pull request suggestions, Kindly check and let me know.

File: packages/actor-extract-links-solid-type-index/lib/ActorExtractLinksSolidTypeIndex.ts
Comment (rubensworks) [inline packages/actor-extract-links-solid-type-index/lib/ActorExtractLinksSolidTypeIndex.ts:121]: This empty line can stay

File: packages/actor-extract-links-solid-type-index/lib/ActorExtractLinksSolidTypeIndex.ts
Comment (rubensworks) [inline packages/actor-extract-links-solid-type-index/lib/ActorExtractLinksSolidTypeIndex.ts:134]: This empty line can stay

File: packages/actor-extract-links-solid-type-index/lib/ActorExtractLinksSolidTypeIndex.ts
Comment (rubensworks) [inline packages/actor-extract-links-solid-type-index/lib/ActorExtractLinksSolidTypeIndex.ts:25]:
```suggestion
  private readonly inference: boolean;
```

File: packages/actor-extract-links-solid-type-index/lib/ActorExtractLinksSolidTypeIndex.ts
Comment (rubensworks) [inline packages/actor-extract-links-solid-type-index/lib/ActorExtractLinksSolidTypeIndex.ts:174]:
```suggestion
      <[IDataSource, ...IDataSource[]]> predicateValue;
```
With perhaps just an if-check before to ensure that length > 0.

File: packages/actor-extract-links-solid-type-index/lib/ActorExtractLinksSolidTypeIndex.ts
Comment (rubensworks) [inline packages/actor-extract-links-solid-type-index/lib/ActorExtractLinksSolidTypeIndex.ts:172]:
```suggestion
  public async fetchPredicateDomains(predicateValue: string[]): Promise<Record<string, string>> {
```

File: packages/actor-extract-links-solid-type-index/lib/ActorExtractLinksSolidTypeIndex.ts
Comment (rubensworks) [inline packages/actor-extract-links-solid-type-index/lib/ActorExtractLinksSolidTypeIndex.ts:257]: This line can go

File: packages/actor-extract-links-solid-type-index/lib/ActorExtractLinksSolidTypeIndex.ts
Comment (rubensworks) [inline packages/actor-extract-links-solid-type-index/lib/ActorExtractLinksSolidTypeIndex.ts:295]:
```suggestion
   * If the domains of query predicates will be considered when checking the type index.
```

File: packages/actor-extract-links-solid-type-index/lib/ActorExtractLinksSolidTypeIndex.ts
Comment (rubensworks) [inline packages/actor-extract-links-solid-type-index/lib/ActorExtractLinksSolidTypeIndex.ts:296]:
```suggestion
   * If false, no predicates will be considered.
```

File: packages/actor-extract-links-solid-type-index/lib/ActorExtractLinksSolidTypeIndex.ts
Comment (rubensworks) [inline packages/actor-extract-links-solid-type-index/lib/ActorExtractLinksSolidTypeIndex.ts:156]: What does this comment mean?

**Review by rubensworks — CHANGES_REQUESTED** [review]
Great work @Sindhu-Vasireddy, this looks a lot clearer already!

Just some small comments here and there. I haven't looked at the unit tests yet, will do so after these comments have been resolved :-)

**Comment by Sindhu-Vasireddy** [comment]
@rubensworks I've now resolved the comments, could you kindly have a look and let me know at your convenience.

File: packages/actor-extract-links-solid-type-index/test/ActorExtractLinksSolidTypeIndex-test.ts
Comment (rubensworks) [inline packages/actor-extract-links-solid-type-index/test/ActorExtractLinksSolidTypeIndex-test.ts:360]: Could we also have a test that builds upon this one, and exposes type index entries for other predicates (like predicate2 and predicate3) as well?
The goal is to check that these entries are ignored by the actor, since we're only using predicate1 in the query.

File: packages/actor-extract-links-solid-type-index/lib/ActorExtractLinksSolidTypeIndex.ts
Comment (rubensworks) [inline packages/actor-extract-links-solid-type-index/lib/ActorExtractLinksSolidTypeIndex.ts:260]: No empty line needed here. (FYI, for these types of issues, it's always a good idea to check your git diff right before or after committing)

File: packages/actor-extract-links-solid-type-index/lib/ActorExtractLinksSolidTypeIndex.ts
Comment (rubensworks) [inline packages/actor-extract-links-solid-type-index/lib/ActorExtractLinksSolidTypeIndex.ts:218]: This line can stay.

File: packages/actor-extract-links-solid-type-index/lib/ActorExtractLinksSolidTypeIndex.ts
Comment (rubensworks) [inline packages/actor-extract-links-solid-type-index/lib/ActorExtractLinksSolidTypeIndex.ts:153]:
```suggestion
  public async linkPredicateDomains(predicateSubjects: Record<string, RDF.Term>,
```
(To remain consistent with the method below)

File: packages/actor-extract-links-solid-type-index/lib/ActorExtractLinksSolidTypeIndex.ts
Comment (rubensworks) [inline packages/actor-extract-links-solid-type-index/lib/ActorExtractLinksSolidTypeIndex.ts:191]: The `?`'s here should not be required, since the query doesn't use OPTIONALs, and will therefore guarantee bindings and values for these variables.

However, I suspect that this was added because otherwise it crashes the unit test.
This is probably because you defined the query engine in such a way that it always gives the same result for the different queries that are being executed.
Perhaps you can modify the unit tests slightly so that it checks which query is being executed (can be done in a hacky manner, by for example just checking for the presence of a certain variable), and returns different results depending on the executed query.

**Review by rubensworks — CHANGES_REQUESTED** [review]
Super!
Mainly some questions about unit tests now.

**Comment by Sindhu-Vasireddy** [comment]
@rubensworks I've now resolved the comments, could you kindly have a look and let me know at your convenience.

File: packages/actor-extract-links-solid-type-index/lib/ActorExtractLinksSolidTypeIndex.ts
Comment (rubensworks) [inline packages/actor-extract-links-solid-type-index/lib/ActorExtractLinksSolidTypeIndex.ts:160]: This linter rule gives you a hint to simplify this code.
Have a look at `Object.fromEntries()`, it might help making this piece of code a bit more compact.

File: packages/actor-extract-links-solid-type-index/lib/ActorExtractLinksSolidTypeIndex.ts
Comment (rubensworks) [inline packages/actor-extract-links-solid-type-index/lib/ActorExtractLinksSolidTypeIndex.ts:186]: This comment is not correct anymore (not an array).

File: packages/actor-extract-links-solid-type-index/lib/ActorExtractLinksSolidTypeIndex.ts
Comment (rubensworks) [reply packages/actor-extract-links-solid-type-index/lib/ActorExtractLinksSolidTypeIndex.ts:276]: The empty is still here :/

File: packages/actor-extract-links-solid-type-index/lib/ActorExtractLinksSolidTypeIndex.ts
Comment (rubensworks) [reply packages/actor-extract-links-solid-type-index/lib/ActorExtractLinksSolidTypeIndex.ts:218]: Bump

File: packages/actor-extract-links-solid-type-index/lib/ActorExtractLinksSolidTypeIndex.ts
Comment (rubensworks) [inline packages/actor-extract-links-solid-type-index/lib/ActorExtractLinksSolidTypeIndex.ts:28]: We should just have one query engine field, so this one can go.

**Review by rubensworks — CHANGES_REQUESTED** [review]
I see you've added `queryEngineLocal`. I suspected you did this to make you tests handle the 2 different types of queries.
While this is a step in the right direction, we should have just one queryEngine field in the actor.

To solve this, we should mock the queryEngine in such a way that we take into account the first argument of `queryBindings` to look at what query we're currently execution (either the type index, or the predicate vocab), and depending on that, return something else.

If this isn't clear, please let me know, and we can setup a meeting where I can show how this can be done.

File: packages/actor-extract-links-solid-type-index/lib/ActorExtractLinksSolidTypeIndex.ts
Comment (rubensworks) [inline packages/actor-extract-links-solid-type-index/lib/ActorExtractLinksSolidTypeIndex.ts:28]: This one can be removed now.

File: packages/actor-extract-links-solid-type-index/lib/ActorExtractLinksSolidTypeIndex.ts
Comment (rubensworks) [inline packages/actor-extract-links-solid-type-index/lib/ActorExtractLinksSolidTypeIndex.ts:193]: Why was this code changed? The previous version looked fine to me. Not sure this one is fully correct.

**Review by rubensworks — CHANGES_REQUESTED** [review]
Tests look perfect now!
Just some finishing touches remaining.

File: packages/actor-extract-links-solid-type-index/lib/ActorExtractLinksSolidTypeIndex.ts
Comment (rubensworks) [inline packages/actor-extract-links-solid-type-index/lib/ActorExtractLinksSolidTypeIndex.ts:196]: I guess this can be simplied now, so that we only return the `domainsArray` (which would fit more with the current method name).
Then above, where we call this method, we can reconstruct the Record that maps predicateValue to this array.

**Review by rubensworks — CHANGES_REQUESTED** [review]
Nice work @Sindhu-Vasireddy !

This should be good to go after my final suggestion.

Before we merge, could you check if this version of the code still works as expected on SolidBench?

**Review by rubensworks — APPROVED** [review]
Code is in perfect shape now.
Really nice work @Sindhu-Vasireddy, this is a really nice contribution!

Once you've confirmed that things still work as intended with SolidBench, I'll merge this.

**Comment by Sindhu-Vasireddy** [comment]
> Code is in perfect shape now. Really nice work @Sindhu-Vasireddy, this is a really nice contribution!
> 
> Once you've confirmed that things still work as intended with SolidBench, I'll merge this.

I have checked that this works with the current version of Solidbench. I have also addressed the changes on that PR as well.

**Comment by rubensworks** [comment]
Perfect, thanks for checking!

Could you just resolve the merge conflict? I'll merge it right away then.

**Comment by rubensworks** [comment]
Really nice work @Sindhu-Vasireddy! 🎉

**Pull request merged**
